### DISCUSSION TITLE
Cluster setup: What do with version controlled files under /var?

### DISCUSSION THREAD

**Discussion by danmichaelo** [pimcore/pimcore#7966] [Q&A / Support] [answered]
Hi!

Planning to move `var` and `web/var` to an NFS share to prepare for a [cluster setup](https://pimcore.com/docs/pimcore/current/Development_Documentation/Installation_and_Upgrade/System_Setup_and_Hosting/Cluster_Setup.html), but the problem is that we currently have a few version controlled folders there: `var/classes`, `var/config` and `var/bundles`. 

Is the solution to move these out of `var/`? At least `var/classes`, which we need to have under version control, since we model our classes locally and re-build them on the servers using [`pimcore:deployment:classes-rebuild`](https://pimcore.com/docs/pimcore/current/Development_Documentation/Deployment/Deployment_Tools.html). Found that I could achieve this by setting`PIMCORE_CLASS_DIRECTORY`.

I'm not so sure about whether we really need to keep `var/config` (`PIMCORE_CONFIGURATION_DIRECTORY`) under version control. It seems like most of the config there [can also be defined in `app/config/pimcore`](https://github.com/pimcore/pimcore/blob/bf116bc2a5ee759c21a02f6cca704a66c6080dd9/lib/Config.php#L104-L107) (`PIMCORE_CUSTOM_CONFIGURATION_DIRECTORY`) instead, which is perhaps better? The one exception I've found so far is [the logo](https://github.com/pimcore/pimcore/blob/4fd1f8d1889b0c5e5f315b6aa1a6dad55fb2e362/bundles/AdminBundle/Controller/Admin/SettingsController.php#L69).

Finally, there's a few bundles that keep stuff under `var/bundles`, for which it seems like there isn't a constant, so this could perhaps be the most difficult part to handle, but I guess I could try to discuss directly with the plugin editors if they could introduce a constant or configuration option to change where they store files.

I'm a little bit 🤷‍♂️ that this turned out more difficult than I first thought, so would love to hear about how others have approached this, and whether you think I'm on the right track or not.

**Comment by dpfaffenbauer** [comment, marked as answer]
Difficult but doable ;).

My setup looks like this:

Shared Files:

```
    '.env',
    'app/config/parameters.yml',
    'app/config/local/staging.yml',
    'var/config/robots.php',
    'var/config/system.php',
    'var/config/debug-mode.php',
    'var/config/maintenance.php',
    'var/config/reports.php',
    'var/config/GeoLite2-City.mmdb',
    'var/bundles/LuceneSearchBundle/state.cnf',
    'var/config/custom-logo.png',
    'var/config/website-settings.php',
```

Shared Directories:

```
    'web/var',
    'var/email',
    'var/recyclebin',
    'var/versions',
    'var/sessions',
    'var/bundles/LuceneSearchBundle/index',
    'var/bundles/LuceneSearchBundle/site-map',
    'var/config/key',
```

Files get symlinked into the release folder and sit on a NFS Mount.

**Reply by danmichaelo** [reply]
Thanks, seems like the way to go is to selectively symlink parts of the `var` folder rather than the whole folder! On my test server, I've now symlinked the following folders, and it seems to work pretty well (except that NFS performance is causing me some pain, but that's nothing to do with Pimcore):

```
  var/application-logger
  var/cache
  var/config
  var/email
  var/logs
  var/pool
  var/recyclebin
  var/tmp
  var/versions
  web/var
```

Decided to try starting out symlinking the whole `config` folder rather than parts of it to reduce management overhead. The only negative I've found so far is that I can't version control the logo, but that's not a huge thing.

**Comment by langert** [comment]
Hi,
pimcore:deployment:classes-rebuild does not create the classes in the database, right? How did you solve this when adding a new object class in pimcore, to get this working with different deployments? We startet to build a structur to first export the clases as json, copy this files to the deployment and import them via ./bin/console pimcore:definition:import:class /class_jsonfile_path.json - but this feels strange... the docs [https://pimcore.com/docs/pimcore/current/Development_Documentation/Deployment/Deployment_Tools.html](url) have no hint to this

**Reply by NiklasBr** [reply]
Try the `create-classes` option:

https://github.com/pimcore/pimcore/blob/7edef74976d037e913a6ea7d2c497a566d62db57/bundles/CoreBundle/Command/ClassesRebuildCommand.php#L42-L47

**Reply by langert** [reply]
I already searched for a command like this, but not as an option - thx!
